src/java/org/apache/cassandra/db/tries/TrieSetCursor.java

```diff
@@ -0,0 +1,144 @@
+        END_START_PREFIX(true, true, false),
+        /// The cursor is positioned at a "point" boundary, i.e. only the descendants of the boundary are covered,
+        /// branches to the left or right are not contained.
+        POINT(false, false, true),
+        /// The cursor is positioned at an end boundary. Branches to the left, as well as descendants of this point are
+        /// covered by the set.
+        END(true, false, true),
+        /// The cursor is positioned at a start boundary. Branches to the right, as well as descendants of this point
+        /// are covered by the set.
+        START(false, true, true),
+        /// The cursor is positioned at a non-effective boundary (an end boundary for the previous range, as well as
+        /// a start for the next). Branches before, after and below this point is covered.
+        COVERED(true, true, true);
+
+        /// Whether the set applied to positions before the cursor's in forward order.
+        final boolean applicableBefore;
+        /// Whether the set applied to positions after the cursor's in forward order.
+        final boolean applicableAfter;
+        /// The state to report as content. This converts prefix states to null to report only the boundaries
+        /// (e.g. for dumping to text).
+        final RangeState asContent;
+
+        RangeState(boolean applicableBefore, boolean applicableAfter, boolean applicableAtPoint)
+        {
+            this.applicableBefore = applicableBefore;
+            this.applicableAfter = applicableAfter;
+            this.asContent = applicableAtPoint ? this : null;
+        }
+
+        /// Whether the positions preceding the current in iteration order are included in the set.
+        public boolean precedingIncluded(Direction direction)
+        {
+            return direction.select(applicableBefore, applicableAfter);
+        }
+
+        /// Whether the descendant branch is fully included in the set.
+        public boolean branchIncluded()
+        {
+            return asContent != null;
+        }
+
+        public RangeState toContent()
+        {
+            return asContent;
+        }
+
+        /// Return an "intersection" state for the combination of two states, i.e. the ranges covered by both states.
+        public RangeState intersect(RangeState other)
+        {
+            return values()[ordinal() & other.ordinal()];
+        }
+
+        /// Return a "union" state for the combination of two states, i.e. the ranges covered by at least one of the states.
+        public RangeState union(RangeState other)
+        {
+            return values()[ordinal() | other.ordinal()];
+        }
+
+        /// Return the "weakly negated" state, i.e. the state that corresponds to flipped areas of coverage to the left
+        /// and right, and the boundary points. See [TrieSet#weakNegation] for more details.
+        public RangeState weakNegation()
+        {
+            return values()[ordinal() ^ 3];
```

Comment:
Maybe write 3 as `0b011` just to show that we flip applicableBefore and applicableAfter ?